### ISSUE TITLE
how to understand the different kernel catching different feature in CNN?

### ISSUE THREAD

**Issue by charles-bu** [ageron/handson-ml#336]
hi, 

I still in learning your book and I moved into the CNN. Though I can know the computing procedure of the CNN, I still was puzzled in the process of using more kernel to catch the different feature. i do understand that one kernel is not sufficient  to catch the picture information. however, once we used to multi kernel, how to let those kernel to catch different feature, just by the different initial setting?

in your book example, you show using two different kernels to catch the vertical and horizontal feature in the forward process. for my understanding, those initial setting may be reset via the BP process. since the target label keeps the same, will it force the those two different kernels' weight converge to the same weight? of course, I do know it will not be happened.  however, how to understand different kernel to catch different feature in the BP process?

Best regard,

Charles

**Comment by charles-bu** [comment]
Hi, 

thank you so much for your prompt replying.

actually, the problem tortured me starting  from the fully connected network and was got my attention during my learning the CNN. let me say in this way, we all do know that in the fully connected network, those neurons in the same layer would catch different data feature respectively. Why does it act in this way? 

after learning your book, I tried to code the fully connected by myself, which gave me more deep understanding of the ANN. meanwhile, I also noticed that the neurons in the same layer were identical, before the initial weights was assigned. it looks those initial weights is the only factor making those neurons to catch different data features and making those neurons different from each other . but why? will it happen two neurons catching the same feature of the data? 

considering a human case, we hold a party with 10 boys and girls. we even could do the same work, assigned the dancer partner for those boys and girls, which is something like our ANN initial weight setting. however, in most cases, most of boys will run for the same girl consciously or unconsciously. why those kind of thing will not happen in the neuron network? just because of the different initial setting? 

Though those ideas is not very much ML or ANN/CNN related, I still wish to have your advice, which is highly appreciated!


Best regards,


Charles

**Comment by ageron** [comment]
Hi Charles,

Good question. The key is to understand that gradient descent tweaks all weights simultaneously, at each iteration. It tries to find the best *combination* of parameters, it does not search for good parameters independently of each other. Clearly a neural network in which all neurons learn different useful things is more powerful than a neural network in which all neurons learn the same thing, so gradient descent will prefer the former. Even if weights are very close to each other at the beginning, gradient descent will tend to push them apart in order to get a better *combination* of neurons.

To go for a human analogy, think of a soccer team. A selector will try to build the best *team*, not just select the best individual players. For example, if he builds a team containing 11 excellent goalkeepers, he will not win the competition. Instead, the best team will have an excellent goalkeeper, excellent defenders, excellent attackers, and so on. Just to be clear, the analogy is that the team is a neural network, each player is a neuron, and the selector is gradient descent.

Let's look at a more concrete example. Suppose your network has just two parameters w1 and w2, and the cost function is the one represented in the following image, with two minima: one around (5, 2), and the other around (2, 5).

<img width="388" alt="screen shot 2018-12-24 at 10 20 51" src="https://user-images.githubusercontent.com/76661/50389476-a43c0900-0765-11e9-92fe-4f9dba76f124.png">

Notice that the cost function is symmetrical, so w1 and w2 are interchangeable, just like neurons in a fully connected layer. After you initialize w1 and w2 randomly, gradient descent will pull the weights towards the closest minimum, so for example if random initialization starts at w1=1 and w2=9, then gradient descent will converge to w1=2 and w2=5. But if you start at w1=9 and w2=1, then gradient descent will converge to w1=5 and w2=2. Note that it is important to initialize the weights randomly, or else gradient descent will not be able to break the symmetry. For example, suppose we initialize the weights at w1=w2=0. Since both minima are equally close to this point, w1 and w2 will be tweaked in exactly the same way, so we will always have w1=w2. After several iterations, gradient descent may end up around w1=w2=3.5, which is NOT a global minimum (not even a minimum, in fact).

I hope this is clear!

Cheers,
Aurélien
